grpc-java/grpc-servlet-jakarta/src/main/java/io/grpc/servlet/jakarta/web/GrpcWebFilter.java

```diff
@@ -0,0 +1,134 @@
+        if (isGrpcWeb(request)) {
+            // wrap the request and response to paper over the grpc-web details
+            GrpcWebHttpResponse wrappedResponse = new GrpcWebHttpResponse(response);
+            HttpServletRequestWrapper wrappedRequest = new HttpServletRequestWrapper(request) {
+                @Override
+                public String getContentType() {
+                    // Adapt the content-type to replace grpc-web with grpc
+                    return super.getContentType().replaceFirst(Pattern.quote(CONTENT_TYPE_GRPC_WEB),
+                            GrpcUtil.CONTENT_TYPE_GRPC);
+                }
+
+                @Override
+                public AsyncContext startAsync() throws IllegalStateException {
+                    return startAsync(this, wrappedResponse);
+                }
+
+                @Override
+                public AsyncContext startAsync(ServletRequest servletRequest, ServletResponse servletResponse)
+                        throws IllegalStateException {
+                    AsyncContext delegate = super.startAsync(servletRequest, servletResponse);
+                    return new DelegatingAsyncContext(delegate) {
+                        @Override
+                        public void complete() {
+                            // Write any trailers out to the output stream as a payload, since grpc-web doesn't
+                            // use proper trailers.
+                            try {
+                                if (wrappedResponse.trailers != null) {
+                                    Map<String, String> map = wrappedResponse.trailers.get();
+                                    if (map != null) {
+                                        // write a payload, even for an empty set of trailers, but not for
+                                        // the absence of trailers.
+                                        int trailerLength = map.entrySet().stream()
+                                                .mapToInt(e -> e.getKey().length() + e.getValue().length() + 4).sum();
+                                        ByteBuffer payload = ByteBuffer.allocate(5 + trailerLength);
+                                        payload.put((byte) 0x80);
+                                        payload.putInt(trailerLength);
+                                        for (Map.Entry<String, String> entry : map.entrySet()) {
+                                            payload.put(entry.getKey().getBytes(StandardCharsets.US_ASCII));
+                                            payload.put(": ".getBytes(StandardCharsets.US_ASCII));
+                                            payload.put(entry.getValue().getBytes(StandardCharsets.US_ASCII));
+                                            payload.put("\r\n".getBytes(StandardCharsets.US_ASCII));
+                                        }
+                                        wrappedResponse.getOutputStream().write(payload.array());
```

Comment:
Might be worth checking that the payload has been fully used (in the case of overflow, we should already get exception)?

Reply:
Unfortunate, write doesn't take a ByteBuffer, just a `byte[]` (or individual `byte`s/etc), so presently I'm just writing the whole array in one go (since the size is known, just using ByteBuffer as a convenient way to append bytes as we go), so we must assume that `OutputStream.write` will throw if it doesn't write enough.

Reply:
I mean, check BB remaining is 0?

Reply:
Ah that seems reasonable yes.

Reply:
Here's the exception that this update will cause, while still doing its best to not break the socket (though it will omit the trailers, as they have null bytes):
```
2022-08-29T21:13:42.749Z | qtp417875774-39      | ERROR | i.g.i.SerializingExecutor | Exception while executing runnable io.grpc.servlet.jakarta.AsyncServletOutputStreamWriter$$Lambda$719/0x0000000800678440@6b16a7fd
java.lang.IllegalStateException: null
        at io.grpc.servlet.jakarta.web.GrpcWebFilter$1$1.complete(GrpcWebFilter.java:83)
        at io.grpc.servlet.jakarta.AsyncServletOutputStreamWriter.lambda$new$1(AsyncServletOutputStreamWriter.java:97)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
        at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
        at io.grpc.internal.SerializingExecutor.schedule(SerializingExecutor.java:102)
        at io.grpc.internal.SerializingExecutor.execute(SerializingExecutor.java:95)
        at io.grpc.servlet.jakarta.ServletServerStream$ServletTransportState.runOnTransportThread(ServletServerStream.java:144)
        at io.grpc.servlet.jakarta.AsyncServletOutputStreamWriter.lambda$new$2(AsyncServletOutputStreamWriter.java:94)
        at io.grpc.servlet.jakarta.AsyncServletOutputStreamWriter.onWritePossible(AsyncServletOutputStreamWriter.java:143)
        at io.grpc.servlet.jakarta.ServletServerStream$GrpcWriteListener.onWritePossible(ServletServerStream.java:218)
        at org.eclipse.jetty.server.HttpOutput.run(HttpOutput.java:1493)
        at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1460)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:606)
        at org.eclipse.jetty.server.HttpChannel.run(HttpChannel.java:457)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:412)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:381)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:268)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:190)
        at org.eclipse.jetty.http2.HTTP2Connection.produce(HTTP2Connection.java:208)
        at org.eclipse.jetty.http2.HTTP2Connection.onFillable(HTTP2Connection.java:155)
        at org.eclipse.jetty.http2.HTTP2Connection$FillableCallback.succeeded(HTTP2Connection.java:378)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
        at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:530)
        at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:379)
        at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:146)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
        at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:412)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:381)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:268)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.lambda$new$0(AdaptiveExecutionStrategy.java:138)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:407)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:894)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1038)
        at java.base/java.lang.Thread.run(Thread.java:829)

```

Tested by negating the new hasRemaining() check, and confirming that the browser no longer thinks more data is on its way, fails with the truncated payload it did get, and the server log contains the stack trace.